server/aesdsocket.c

```diff
@@ -18,178 +19,402 @@
 #define FILE_PATH "/var/tmp/aesdsocketdata"
 #define CONNECTIONS 20
 
+
+// Thread elements in structure
+typedef struct {                    
+    pthread_t   thread_id;                
+    bool        conn_build;
+    int         conn_socket;
+} thread_info_t;
+
 pthread_mutex_t mutex;
+pthread_t timestamp_threadId;
+
+
+// Singly linked list node to hold thread information
+typedef struct slist_data_s slist_data_t;
+struct slist_data_s
+{
+    thread_info_t* threadInfo;
+    SLIST_ENTRY(slist_data_s) entries;
+};
+
+SLIST_HEAD(slisthead, slist_data_s) head;
+struct addrinfo* addrinfo_s;
 
 // Global variables
 int sockfd = -1;
-FILE* fp = NULL;
-struct addrinfo* addrinfo_s;
+FILE* fp = 0;
+unsigned int sendSizeTotal = 0;
 char* conn_ip;
+bool callTimeStampthread = false;
+
+// Signal handler for cleaning up child process
+static void signal_handler_child(int signal_number)
+{
+    if (signal_number == SIGCHLD)
+    {
+        int saved_errno = errno;
+        syslog(LOG_ERR, "In signal_handler_child");
+        while (waitpid((pid_t)(-1), 0, WNOHANG) > 0) {}
+        errno = saved_errno;
+    }
+}
+
+
+// Function to clean up the linked list and join threads
+static void cleaup_slist(void)
+{
+    slist_data_t *datap;
+    while (!SLIST_EMPTY(&head))
+    {
+        datap = SLIST_FIRST(&head);
+        if (datap->threadInfo && datap->threadInfo->conn_socket > 0)
+        {
+         	close(datap->threadInfo->conn_socket);
+        	datap->threadInfo->conn_socket = 0;
+        	pthread_join(datap->threadInfo->thread_id, NULL);
+        	free(datap->threadInfo);
+        }
+        SLIST_REMOVE_HEAD(&head, entries);
+        free(datap);
+    }
+}
+
 
 // Signal handler for SIGINT and SIGTERM
 static void signal_handler(int signal_number)
 {
     if (signal_number == SIGINT || signal_number == SIGTERM)
     {
-        syslog(LOG_DEBUG, "Caught signal, exiting");
-        
-        // Cleanup and close
-        if (sockfd != -1) 
+        syslog(LOG_ERR, "Caught signal \n");
+       
+       // Free allocated resources and cleanup
+        if (addrinfo_s)
         {
-        	close(sockfd);
+         	freeaddrinfo(addrinfo_s);
+                addrinfo_s = 0;
         }
         
-        if (fp) 
+        if (fp)
         {
-        	fclose(fp);
         	remove(FILE_PATH);
-	}
-        if (addrinfo_s)
-        { 
-        	freeaddrinfo(addrinfo_s);
+        	fclose(fp);
         }
-
-        pthread_mutex_destroy(&mutex);
-        closelog();
+        cleaup_slist();
```

Comment:
Cleaning up here will mean threads will not be cleaned up until the server exits.  Would be better to cleanup in the while() loop which adds new threads.